core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java

```diff
@@ -71,49 +74,109 @@ public Endpoint select(ClientRequestContext ctx) {
         private static final class EndpointsAndWeights {
             private final List<Endpoint> endpoints;
             private final boolean weighted;
-            private final int maxWeight;
-            private final int totalWeight;
+            private final long totalWeight; // prevent overflow by using long
+
+            private static final class EndpointsGroupByWeight {
+                int startIndex;
+                int weight;
+                long accumulatedWeight;
+
+                EndpointsGroupByWeight(int startIndex, int weight, long accumulatedWeight) {
+                    this.startIndex = startIndex;
+                    this.weight = weight;
+                    this.accumulatedWeight = accumulatedWeight;
+                }
+            }
+
+            private final EndpointsGroupByWeight[] endpointsGroupByWeight;
 
             EndpointsAndWeights(Iterable<Endpoint> endpoints) {
                 int minWeight = Integer.MAX_VALUE;
                 int maxWeight = Integer.MIN_VALUE;
-                int totalWeight = 0;
+                long totalWeight = 0;
+
+                // get min and max weight
                 for (Endpoint endpoint : endpoints) {
                     final int weight = endpoint.weight();
                     minWeight = Math.min(minWeight, weight);
                     maxWeight = Math.max(maxWeight, weight);
-                    totalWeight += weight;
                 }
-                this.endpoints = ImmutableList.copyOf(endpoints);
-                this.maxWeight = maxWeight;
+
+                // prepare endpoints
+                List<Endpoint> endps = ImmutableList.copyOf(endpoints)
+                        .stream()
+                        .filter(endpoint -> endpoint.weight() > 0) // only process endpoint with weight > 0
+                        .sorted(Comparator
+                                .comparing(Endpoint::weight)
+                                .thenComparing(Endpoint::host)
+                                .thenComparingInt(Endpoint::port))
+                        .collect(Collectors.toList());
+                int numEndpoints = endps.size();
+
+                // accumulation
+                LinkedList<EndpointsGroupByWeight> accumulatedGroups = new LinkedList<>();
+                EndpointsGroupByWeight currentGroup = null;
+                int rest = numEndpoints;
```

Comment:
`long rest`?

Reply:
Number of Endpoints is int (in signature of size() of List too). So I think int is ok.

Reply:
Yeah but you only use it by casting to `long` - if you use `long` here, I think you can remove all the casts.

Reply:
Yeah, and I need to change type of `startIndex` to `long` too to remove other casts.